Source/WebCore/editing/Editing.cpp

```diff
@@ -1222,4 +1223,117 @@ HashSet<RefPtr<HTMLImageElement>> visibleImageElementsInRangeWithNonLoadedImages
     return result;
 }
 
+enum class RangeEndpoint : bool { Start, End };
+static std::optional<BoundaryPoint> visuallyContiguousBoundary(InlineBoxAndOffset&& fromPosition, InlineBoxAndOffset&& toPosition, TextDirection lineDirection, RangeEndpoint startOrEnd)
+{
+    struct VisuallyAdjacentTextRange {
+        Ref<Text> text;
+        unsigned startOffset { 0 };
+        unsigned endOffset { 0 };
+        unsigned minOffset { 0 };
+        unsigned maxOffset { 0 };
+        TextDirection direction { TextDirection::LTR };
+
+        unsigned totalLength() const { return maxOffset > minOffset ? maxOffset - minOffset : 0; }
+        unsigned visuallySelectedLength() const { return endOffset > startOffset ? endOffset - startOffset : 0; }
+
+        BoundaryPoint startBoundary() const { return { text, minOffset }; }
+        BoundaryPoint endBoundary() const { return { text, maxOffset }; }
+    };
+
+    auto [fromBox, fromOffset] = fromPosition;
+    auto [toBox, toOffset] = toPosition;
+    if (fromBox == toBox)
+        return std::nullopt;
+
+    bool isSelectionStart = startOrEnd == RangeEndpoint::Start;
+    auto advance = [&](InlineIterator::BoxIterator box) {
+        return isSelectionStart == (lineDirection == TextDirection::LTR) ? box->nextOnLine() : box->previousOnLine();
+    };
+
+    Vector<VisuallyAdjacentTextRange> ranges;
+    for (auto box = fromBox; box; box = advance(box)) {
```

Comment:
So a while ago, I added RenderedPosition to deal with walk inline boxes accounting for bidi-levels.
Maybe we can use it here?

Reply:
Good call — yes, I think I'll need to adopt `RenderedPosition` here, especially so that I can achieve some sensible behavior with respect to nested bidi. Will try out a few things and update the PR — thanks!